File: internal/connector/tegrauefi/efivar.go
package tegrauefi

// efivarfs access for the NVIDIA RootfsStatusSlot variables.
//
// Validated variable format (t234/r36, t264/r38, and Orin Nano t234/r39.2 —
// the last read directly off the device for WDY-1742, identical 8-byte layout):
//   bytes 0..3  EFI attributes, 0x07 = NV+BS+RT (little-endian uint32)
//   bytes 4..7  status (little-endian uint32): 0 = normal, 0xFF = unbootable
//
// Writing attrs(0x07) + status(0) in a single 8-byte write resets a slot
// to normal AND re-seeds the firmware retry budget (observed on Thor:
// retry_count back to 3). efivarfs marks variables immutable by default,
// so the immutable inode flag must be cleared before writing — the
// chattr -i equivalent, done natively via FS_IOC_GETFLAGS/SETFLAGS.

import (
	"fmt"
	"os"

	"golang.org/x/sys/unix"
)

// statusNormal is the full 8-byte payload that rehabilitates a slot.
var statusNormal = []byte{0x07, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00}

// fsImmutableFL is FS_IMMUTABLE_FL from linux/fs.h (not exported by
// golang.org/x/sys/unix).
const fsImmutableFL = 0x00000010

// readStatus returns the raw variable content. A valid status var is
// exactly 8 bytes; callers check size themselves (the JP6 incident
// involved a wrong-sized variable).
func readStatus(path string) ([]byte, error) {
	return os.ReadFile(path)
}

// statusIsWellFormed reports whether raw is the validated 8-byte
// attrs+status layout. Confirmed identical on t234/r36, t264/r38, and Orin
// Nano t234/r39.2 (read off the device for WDY-1742). A different size is a
// layout we have NOT validated on this board: BootIsCompromised treats it as
// inconclusive rather than compromised, so the documented JP6 wrong-sized
// variable cannot force a false rollback. The engine's running-slot vs
// target-slot check and the ESRT cascade remain the authoritative guards.
func statusIsWellFormed(raw []byte) bool {
	return len(raw) == 8
}

// statusIsNormal reports whether raw content is a well-formed
// "normal" status variable (size 8, all four status bytes zero).
func statusIsNormal(raw []byte) bool {
	if !statusIsWellFormed(raw) {
		return false
	}
	return raw[4] == 0 && raw[5] == 0 && raw[6] == 0 && raw[7] == 0
}

// clearImmutable removes the immutable inode flag (efivarfs sets it by
// default). Filesystems without flag support (e.g. tmpfs in tests)
// return ENOTTY/EOPNOTSUPP — treated as "nothing to clear".
func clearImmutable(path string) error {
	f, err := os.OpenFile(path, os.O_RDONLY, 0)
	if err != nil {
		return err
	}
	defer f.Close()

	flags, err := unix.IoctlGetInt(int(f.Fd()), unix.FS_IOC_GETFLAGS)
	if err != nil {
		if err == unix.ENOTTY || err == unix.EOPNOTSUPP || err == unix.ENOSYS {
			return nil
		}
		return fmt.Errorf("get inode flags: %w", err)
	}
	if flags&fsImmutableFL == 0 {
		return nil
	}
	flags &^= fsImmutableFL
	if err := unix.IoctlSetPointerInt(int(f.Fd()), unix.FS_IOC_SETFLAGS, flags); err != nil {
		return fmt.Errorf("clear immutable flag: %w", err)
	}
	return nil
}

// writeStatusNormal writes the 8-byte normal payload in a single
// write(2), matching the validated dd pattern, then verifies read-back.
func writeStatusNormal(path string) error {
	if err := writeVar(path, statusNormal); err != nil {
		return fmt.Errorf("status var: %w", err)
	}
	raw, err := readStatus(path)
	if err != nil {
		return fmt.Errorf("read-back: %w", err)
	}
	if !statusIsNormal(raw) {
		return fmt.Errorf("read-back mismatch: % x", raw)
	}
	return nil
}

// writeVar writes a complete efivar payload (attrs + data) in a single
// write(2), clearing the immutable flag first if the variable exists.
func writeVar(path string, payload []byte) error {
	if _, err := os.Stat(path); err == nil {
		if err := clearImmutable(path); err != nil {
			return err
		}
	}
	f, err := os.OpenFile(path, os.O_WRONLY|os.O_CREATE, 0o644)
	if err != nil {
		return err
	}
	if _, err := f.Write(payload); err != nil {
		f.Close()
		return fmt.Errorf("write: %w", err)
	}
	if err := f.Close(); err != nil {
		return fmt.Errorf("close: %w", err)
	}
	return nil
}

// osIndicationsProcessCapsule is bit 2 of the OsIndications UINT64:
// "process capsule(s) on next boot". Validated on Thor: armed variable
// reads 07 00 00 00 04 00 00 00 00 00 00 00 (4-byte attrs + UINT64).
//
// The SAME bit in the companion OsIndicationsSupported variable is the
// firmware's capability signal: it means FILE_CAPSULE_DELIVERY (capsule-on-disk)
// is supported. Both variables share the 4-byte-attrs + UINT64 layout, so
// byte[4] carries bits 0..7.
const osIndicationsProcessCapsule = 0x04

// firmwareSupportsCapsuleOnDisk reports whether the firmware advertises
// FILE_CAPSULE_DELIVERY in OsIndicationsSupported. Absent, unreadable, or a
// short variable → not supported. Verified on-device: Orin Nano t234/r39.2
// reads 06 00 00 00 45 ... (bit 2 set) and Thor t264/r38 likewise.
func firmwareSupportsCapsuleOnDisk(path string) bool {
	raw, err := os.ReadFile(path)
	if err != nil || len(raw) < 5 {
		return false
	}
	return raw[4]&osIndicationsProcessCapsule != 0
}

// clearOsIndicationsCapsuleBit disarms capsule processing (rollback
// before reboot). Preserves other bits; a missing variable is a no-op.
func clearOsIndicationsCapsuleBit(path string) error {
	raw, err := os.ReadFile(path)
	if os.IsNotExist(err) {
		return nil
	}
	if err != nil {
		return fmt.Errorf("OsIndications: %w", err)
	}
	if len(raw) < 5 || raw[4]&osIndicationsProcessCapsule == 0 {
		return nil // bit not set
	}
	var value uint64
	for i := 0; i < 8 && 4+i < len(raw); i++ {
		value |= uint64(raw[4+i]) << (8 * i)
	}
	value &^= osIndicationsProcessCapsule

	payload := make([]byte, 12)
	payload[0] = 0x07
	for i := 0; i < 8; i++ {
		payload[4+i] = byte(value >> (8 * i))
	}
	if err := writeVar(path, payload); err != nil {
		return fmt.Errorf("OsIndications: %w", err)
	}
	return nil
}

// setOsIndicationsCapsuleBit sets bit 2, preserving any other bits the
// variable already carries, then verifies the bit reads back. Port of
// oe4t-set-uefi-OSIndications + switch-rootfs verify_osindications.
func setOsIndicationsCapsuleBit(path string) error {
	var value uint64
	if raw, err := os.ReadFile(path); err == nil && len(raw) >= 5 {
		for i := 0; i < 8 && 4+i < len(raw); i++ {
			value |= uint64(raw[4+i]) << (8 * i)
		}
	} // absent or malformed: start from 0, the write creates it

	value |= osIndicationsProcessCapsule

	payload := make([]byte, 12)
	payload[0] = 0x07 // NV+BS+RT
	for i := 0; i < 8; i++ {
		payload[4+i] = byte(value >> (8 * i))
	}
	if err := writeVar(path, payload); err != nil {
		return fmt.Errorf("OsIndications: %w", err)
	}

	raw, err := os.ReadFile(path)
	if err != nil {
		return fmt.Errorf("OsIndications read-back: %w", err)
	}
	if len(raw) < 5 || raw[4]&osIndicationsProcessCapsule == 0 {
		return fmt.Errorf("OsIndications read-back: capsule bit not set (% x)", raw)
	}
	return nil
}

File: swift/Sources/TegraUEFI/EfiVar.swift
import CLIError
import Glibc
import LinuxSys

// efivarfs access for the NVIDIA RootfsStatusSlot variables.
//
// Ports the RootfsStatusSlot-primitive half of
// `internal/connector/tegrauefi/efivar.go` (`readStatus`,
// `statusIsWellFormed`, `statusIsNormal`, `clearImmutable`,
// `writeStatusNormal`, `writeVar`). That file's OsIndications/capsule-bit
// helpers are ported by a later task, alongside the rest of the
// `Connector` conformance (Tasks 8.2-8.4).
//
// Validated variable format (t234/r36, t264/r38, and Orin Nano
// t234/r39.2 — the last read directly off the device for WDY-1742,
// identical 8-byte layout):
//   bytes 0..3  EFI attributes, 0x07 = NV+BS+RT (little-endian uint32)
//   bytes 4..7  status (little-endian uint32): 0 = normal, 0xFF = unbootable
//
// Writing attrs(0x07) + status(0) in a single 8-byte write resets a slot
// to normal AND re-seeds the firmware retry budget (observed on Thor:
// retry_count back to 3). efivarfs marks variables immutable by default,
// so the immutable inode flag must be cleared before writing — the
// `chattr -i` equivalent, via `LinuxSys.setImmutable`.

/// Errors from the `EfiVar` primitive. All are fatal to whatever
/// slot-status operation triggered them; there is no case where the
/// caller can usefully continue without the variable in a known state.
public enum EfiVarError: Error, Equatable, ExitCoded {
    /// Reading the variable's raw bytes failed.
    case readStatus(String)
    /// Clearing the immutable inode flag failed for a reason other than
    /// "this filesystem doesn't support the flag".
    case clearImmutable(String)
    /// The single `write(2)` of the variable's payload failed.
    case write(String)
    /// `writeStatusNormal` wrote the reset payload but the read-back
    /// afterward did not come back as a well-formed "normal" status.
    case readBackMismatch([UInt8])

    /// All `EfiVar` failures are fatal to the in-progress slot operation.
    public var exitCode: Int32 { 1 }
}

/// Free functions over the NVIDIA RootfsStatusSlot efivar layout. Every
/// function takes the variable's full path as a parameter (rather than
/// hard-coding `efivarsDir`), which is what makes them testable against a
/// plain temp file — production callers (the `Connector` conformance in a
/// later task) build the real path against `efivarsDir` themselves.
public enum EfiVar {
    /// Real efivarfs mountpoint the Tegra RootfsStatusSlot variables live
    /// under. Exposed as a default for production callers; every function
    /// below takes an explicit `path` instead of reading this directly, so
    /// tests can point at an arbitrary temp file.
    public static let efivarsDir = "/sys/firmware/efi/efivars"

    /// The full 8-byte payload that rehabilitates a slot: attrs `0x07`
    /// (NV+BS+RT) + status `0` (normal).
    public static let statusNormal: [UInt8] = [0x07, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00]

    /// `FS_IOC_GETFLAGS`/`FS_IOC_SETFLAGS` failures meaning "this
    /// filesystem doesn't support the immutable inode flag ioctl at
    /// all" — nothing to clear. Matches `efivar.go`'s `clearImmutable`
    /// errno allowlist exactly.
    private static let unsupportedImmutableErrnos: Set<Int32> = [ENOTTY, EOPNOTSUPP, ENOSYS]

    /// Returns the raw variable content via a whole-file read. A valid
    /// status var is exactly 8 bytes; callers check size themselves via
    /// `statusIsWellFormed`/`statusIsNormal` (the JP6 incident involved a
    /// wrong-sized variable).
    public static func readStatus(_ path: String) throws -> [UInt8] {
        let fd: Int32
        do {
            fd = try LinuxSys.openRead(path)
        } catch {
            throw EfiVarError.readStatus("open \(path): \(error)")
        }
        defer { LinuxSys.close(fd) }

        var out: [UInt8] = []
        var chunk = [UInt8](repeating: 0, count: 4096)
        while true {
            let n: Int
            do {
                n = try chunk.withUnsafeMutableBytes { buf in
                    try LinuxSys.read(fd, buf)
                }
            } catch {
                throw EfiVarError.readStatus("read \(path): \(error)")
            }
            if n == 0 { break }
            out.append(contentsOf: chunk[0..<n])
        }
        return out
    }

    /// Reports whether `raw` is the validated 8-byte attrs+status layout.
    /// Confirmed identical on t234/r36, t264/r38, and Orin Nano t234/r39.2
    /// (read off the device for WDY-1742). A different size is a layout we
    /// have NOT validated on this board: `bootIsCompromised` (a later
    /// task) treats it as inconclusive rather than compromised, so the
    /// documented JP6 wrong-sized variable cannot force a false rollback.
    public static func statusIsWellFormed(_ raw: [UInt8]) -> Bool {
        raw.count == 8
    }

    /// Reports whether `raw` is a well-formed "normal" status variable
    /// (size 8, all four status bytes zero).
    public static func statusIsNormal(_ raw: [UInt8]) -> Bool {
        guard statusIsWellFormed(raw) else { return false }
        return raw[4] == 0 && raw[5] == 0 && raw[6] == 0 && raw[7] == 0
    }

    /// Removes the immutable inode flag (efivarfs sets it by default) via
    /// `LinuxSys.setImmutable(path, false)`. Filesystems without flag
    /// support (e.g. tmpfs in tests) fail with `ENOTTY`/`EOPNOTSUPP`/
    /// `ENOSYS` — treated as "nothing to clear", matching `efivar.go`.
    /// Any other failure (e.g. missing `CAP_LINUX_IMMUTABLE` for an
    /// actual flag change) propagates as `EfiVarError.clearImmutable`.
    public static func clearImmutable(_ path: String) throws {
        do {
            try LinuxSys.setImmutable(path, false)
        } catch let error as SysError {
            if unsupportedImmutableErrnos.contains(error.errno) {
                return
            }
            throw EfiVarError.clearImmutable("\(path): \(error)")
        }
    }

    /// Writes a complete efivar payload (attrs + data) in a single
    /// `write(2)`, clearing the immutable flag first if the variable
    /// already exists. Ports `efivar.go`'s `writeVar` exactly, except it
    /// never creates the variable (`LinuxSys.openWriteExisting` — unlike
    /// Go's `os.O_CREATE` — deliberately fails loudly on a missing path
    /// rather than papering over it with a plain file; a missing
    /// firmware-owned efivar is a configuration error, not something to
    /// silently fix up).
    public static func writeVar(_ path: String, _ payload: [UInt8]) throws {
        if pathExists(path) {
            try clearImmutable(path)
        }

        let fd: Int32
        do {
            fd = try LinuxSys.openWriteExisting(path)
        } catch {
            throw EfiVarError.write("open \(path): \(error)")
        }

        do {
            try payload.withUnsafeBytes { buf in
                _ = try LinuxSys.write(fd, buf)
            }
        } catch {
            LinuxSys.close(fd)
            throw EfiVarError.write("write \(path): \(error)")
        }
        LinuxSys.close(fd)
    }

    /// Writes the 8-byte normal payload in a single `write(2)`, matching
    /// the validated `dd` pattern, then verifies the read-back.
    public static func writeStatusNormal(_ path: String) throws {
        do {
            try writeVar(path, statusNormal)
        } catch {
            throw EfiVarError.write("status var: \(error)")
        }

        let raw: [UInt8]
        do {
            raw = try readStatus(path)
        } catch {
            throw EfiVarError.readStatus("read-back: \(error)")
        }
        guard statusIsNormal(raw) else {
            throw EfiVarError.readBackMismatch(raw)
        }
    }

    /// Reports whether `path` exists, mirroring the `os.Stat(path); err
    /// == nil` guard in `efivar.go`'s `writeVar` — clearing the immutable
    /// flag is only attempted on a variable that's actually there.
    private static func pathExists(_ path: String) -> Bool {
        path.withCString { Glibc.access($0, F_OK) == 0 }
    }
}
